File: Cisco/1FriendsOfAppropriateAges.py
from collections import Counter


class Solution(object):
    def numFriendRequests(self, ages):
        """
        :type ages: List[int]
        :rtype: int
        """
        age_groups = Counter(
            ages
        )  # O(n), but cardinality of `age_groups` map is max. 120, so space is O(1)
        num_requests = 0

        for age_X, num_X in age_groups.items():  # O(120) ~ O(1)
            for age_Y, num_Y in age_groups.items():  # O(120) ~ O(1)
                if age_Y > 0.5 * age_X + 7 and age_Y <= age_X:
                    num_requests += num_X * num_Y
                    if age_X == age_Y:
                        num_requests -= num_X

        return num_requests

File: Cisco/LinkedInList/C++/1FriendsOfAppropriateAges.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int numFriendRequests(vector<int> &ages)
    {
        vector<int> ageGroups(121, 0);
        for (auto &age : ages) // O(N)
            ageGroups[age] += 1;

        int requestCount = 0;
        for (int x = 0; x < ageGroups.size(); ++x) // O(1)
            for (int y = 0; y < ageGroups.size(); ++y) // O(1)
            {
                if (y <= x && y > 0.5 * x + 7)
                {

                    requestCount += ageGroups[x] * ageGroups[y];
                    if (x == y)
                        requestCount -= ageGroups[x];
                }
            }

        return requestCount;
    }
};